### PR TITLE
Revert python build isolation & setuptools source install

### PR DESCRIPTION
**Pull request by haampie**
- Revert "Improve build isolation in PythonPipBuilder (#40224)"
- Revert "py-setuptools: sdist + rpath patch backport (#40205)"

The venv for build isolation is causing issues on Debian due to a different install layout of packages when Python is run from a venv vs from the underlying system executable: PYTHONPATH at runtime is computed incorrectly.

Current decision is to revert this and go back to the old broken state of things for 0.21, and get a proper fix in after the release, as outlined in #40773.

### CODE REVIEW

File: lib/spack/spack/build_systems/python.py [hunk 1]
```diff
@@ -511,36 +464,8 @@ def global_options(self, spec, prefix):
...
-            # There are no environment variables to disable the system site-packages, so we use a
-            # virtual environment instead. The downside of this approach is that pip produces
-            # incorrect shebangs that refer to the virtual environment, which we have to fix up.
-            python("-m", "venv", "--without-pip", self._build_venv_path)
-            pip = self._build_venv_python
-        else:
-            # For a Spack managed Python, system site-packages is empty/unused by design, so it
-            # suffices to disable user site-packages, for which there is an environment variable.
-            pip = python
-            pip.add_default_env("PYTHONNOUSERSITE", "1")
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/python.py:541]: Can we keep this? I think this is the part that found bugs in our existing packages

File: var/spack/repos/builtin/packages/py-pip/package.py [hunk 2]
```diff
@@ -88,9 +88,6 @@ class PyPip(Package, PythonExtension):
     extends("python")
     depends_on("python@3.7:", when="@22:", type=("build", "run"))
 
-    # Uses collections.MutableMapping
-    depends_on("python@:3.9", when="@:19.1", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pip/package.py]: We should def keep this

File: var/spack/repos/builtin/packages/py-setuptools/package.py [hunk 3]
```diff
@@ -69,15 +200,27 @@ def url_for_version(self, version):
...
-    depends_on("py-pip@:18", when="@:53", type="build")
+    # https://github.com/pypa/setuptools/issues/3661
+    conflicts("python@3.12:", when="@:67")
 
-    # Uses HTMLParser.unescape
-    depends_on("python@:3.8", when="@:41.0", type=("build", "run"))
+    depends_on("py-pip", type="build")
 
-    # Uses collections.MutableMapping
-    depends_on("python@:3.9", when="@:40.4.2", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-setuptools/package.py]: Also worth keeping this

Comment (haampie) [reply var/spack/repos/builtin/packages/py-setuptools/package.py]: Did you mean this upperbound specifically, or the rest too?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-setuptools/package.py]: Just these two lines with the comment and upper bound

File: lib/spack/spack/build_systems/python.py [hunk 4]
```diff
@@ -556,31 +481,11 @@ def install(self, pkg, spec, prefix):
         else:
             args.append(".")
 
+        pip = spec["python"].command
+        pip.add_default_env("PYTHONNOUSERSITE", "1")
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/python.py:489]: Can we keep a comment explaining what this is? I won't remember a year from now

File: var/spack/repos/builtin/packages/py-setuptools/package.py [hunk 5]
```diff
@@ -69,15 +200,30 @@ def url_for_version(self, version):
     depends_on("python@2.7:2.8,3.5:", when="@44", type=("build", "run"))
     depends_on("python@2.7:2.8,3.4:", when="@:43", type=("build", "run"))
 
-    # Newer pip requires setuptools to be installed, before building
-    # setuptools. This issue was fixed or worked around in setuptools 54+
-    depends_on("py-pip@:18", when="@:53", type="build")
-
-    # Uses HTMLParser.unescape
-    depends_on("python@:3.8", when="@:41.0", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-setuptools/package.py]: We should keep this too

Comment (haampie) [reply var/spack/repos/builtin/packages/py-setuptools/package.py]: I asked...

File: lib/spack/spack/build_systems/python.py [hunk 6]
```diff
@@ -556,31 +481,15 @@ def install(self, pkg, spec, prefix):
         else:
             args.append(".")
 
+        pip = spec["python"].command
+        # Hide user packages, since we don't have build isolation. This is
+        # necessary because pip / setuptools may run hooks from arbitrary
+        # packages during the build. There is no equivalent variable to hide
+        # system packages, so we this is not reliable for external Python.
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/python.py]:
```suggestion
        # system packages, so this is not reliable for external Python.
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**